### ISSUE TITLE
Panics when using a cloned session from within closure.

### ISSUE THREAD

**Issue by glademiller** [tokio-rs/tokio-curl#10]
I've found it surprisingly easy to get this library to panic but I'm probably doing something wrong as I'm new to Futures in Rust. I've had several examples but the latest is. 

```rust
    let mut lp = Core::new().unwrap();
    let session = Session::new(lp.handle());
    // let session2 = session.clone();
    let mut req = Easy::new();
    req.custom_request("GET").unwrap();
    req.url("https://www.rust-lang.org").unwrap();
    let res = session.perform(req)
        .then(|resp| {
            let status_code = resp.ok().and_then(|mut v| v.response_code().ok()).unwrap_or(0);
            let mut req2 = Easy::new();
            req2.custom_request("GET").unwrap();
            req2.url("https://www.rust-lang.org").unwrap();
            let new_session = session.clone();
            new_session.perform(req2)
        })
        .map(|mut resp2| {
            let status_code = resp2.response_code().unwrap_or(0);
            status_code
        });
    let result = lp.run(res);
    println!("{:?}", result);
```

if I use session2 that was cloned outside the closure everything is fine. My best guess is I'm misunderstanding something fundamental any help is appreciated though.

**Comment by glademiller** [comment]
Updated formatting

**Comment by alexcrichton** [comment]
Oh dear this sounds bad! And definitely unintentional :)

I unfortunately can't reproduce this locally though with [this code](https://gist.github.com/alexcrichton/21ac0976d11254992615894511178955). Could you gist the panic that you're seeing as well? It'd also be helpful to set `RUST_BACKTRACE` to see what's calling what.

**Comment by glademiller** [comment]
Ok I narrowed it down a bit more and I did leave out the detail that this is on Windows using nightly

The panic is actually after the requests are made and the result is printed. Below is the full output when run with RUST_BACKTRACE
```
Ok(200)
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libcore\result.rs:837
stack backtrace:
   0:     0x7ff6da7c377a - std::panicking::default_hook::{{closure}}
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libstd\panicking.rs:354
   1:     0x7ff6da7c2c86 - std::panicking::default_hook
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libstd\panicking.rs:365
   2:     0x7ff6da7c665d - std::panicking::rust_panic_with_hook
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libstd\panicking.rs:553
   3:     0x7ff6da7c64a8 - std::panicking::begin_panic<collections::string::String>
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libstd\panicking.rs:515
   4:     0x7ff6da7c6414 - std::panicking::begin_panic_fmt
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libstd\panicking.rs:499
   5:     0x7ff6da7c63a9 - std::panicking::rust_begin_panic
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libstd\panicking.rs:475
   6:     0x7ff6da7c82a7 - core::panicking::panic_fmt
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libcore\panicking.rs:69
   7:     0x7ff6da74694b - core::result::unwrap_failed<std::sync::mpsc::SendError<tokio_curl::imp::Message>>
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libcore\macros.rs:29
   8:     0x7ff6da73aa61 - core::result::Result<(), std::sync::mpsc::SendError<tokio_curl::imp::Message>>::unwrap<(),std::sync::mpsc::SendError<tokio_curl::imp::Message>>
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libcore\result.rs:737
   9:     0x7ff6da75a9fe - tokio_curl::imp::Sender<tokio_curl::imp::Message>::send<tokio_curl::imp::Message>
                        at C:\Users\glade.miller\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-curl-0.1.4\src\windows.rs:220
  10:     0x7ff6da75817a - tokio_curl::imp::{{impl}}::drop
                        at C:\Users\glade.miller\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-curl-0.1.4\src\windows.rs:88
  11:     0x7ff6da72265a - core::result::unwrap_failed<curl::error::Error>
  12:     0x7ff6da721cee - core::result::unwrap_failed<curl::error::Error>
  13:     0x7ff6da72d02f - dynamics_enrollment_worker_rs::main
                        at C:\Users\glade.miller\Documents\code\dynamics-enrollment-worker-rs\src\main.rs:82
  14:     0x7ff6da7c7231 - panic_unwind::__rust_maybe_catch_panic
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libpanic_unwind\lib.rs:98
  15:     0x7ff6da7c691a - std::rt::lang_start
                        at C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\libstd\rt.rs:51
  16:     0x7ff6da72d5db - main
  17:     0x7ff6da80d05f - __scrt_common_main_seh
                        at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:255
  18:     0x7ffd7a5d8363 - BaseThreadInitThunk
error: process didn't exit successfully: `target\debug\dynamics-enrollment-worker-rs.exe` (exit code: 101)
```

**Comment by glademiller** [comment]
Looking at the stack trace myself it seems like this unwrap is the culprit. https://github.com/tokio-rs/tokio-curl/blob/master/src/windows.rs#L220

**Comment by alexcrichton** [comment]
Interesting! That getting hit probably actually means that the worker thread here panicked first. Did you see any other panic logs come up when testing?

**Comment by glademiller** [comment]
No the only output from the program is in the comment above.

**Comment by glademiller** [comment]
The line referenced in main.rs is the closing bracket of the  main function.

**Comment by alexcrichton** [comment]
Aha! Ok I was able to reproduce on Windows and I was actually hitting a segfault. I addressed that in https://github.com/tokio-rs/tokio-curl/commit/78f6d5d933eb9ffee58e45177710e466673be9ba though which I think may fix this bug as well? Could you test out that commit and see if it works for you?

**Comment by glademiller** [comment]
That rev does indeed fix the issue thanks!

**Comment by alexcrichton** [comment]
Great! I've published 0.1.5 with that fix.
